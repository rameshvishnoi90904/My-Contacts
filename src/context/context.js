import React, { useReducer, useMemo } from 'react';
import PropTypes from 'prop-types';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { sortUsers } from '../helper';

const initialState = {
    contactList: [],
    syncEnabled: false
};

const initialContext = [{ ...initialState }, () => {}];

export const UserContext = React.createContext(initialContext);

/**
 * 
 * @constructor
 * @param {Object} state - Old State
 * @param {Object} update - object containing type and payload for updating state
 * 
 * @returns {Object} - updated state value
 */
const updater = (state, update) => {
    var toUpdate = {...state};
    switch (update.type) {
        case 'init': {
            const sortedArray = sortUsers([...state.contactList,...update.data.contactList])
            toUpdate = {
                ...state,
                contactList: sortedArray
            }
            break;
        }
            
        case 'addNewContact': {
            // check if contact already present with same recordid
            const toAdd = update.data.contactData;
            const index = state.contactList.findIndex((item) => {
                return (item.recordID == toAdd.recordID)
            });
            let updateContactList;
            if (index == -1) {
                updateContactList = [...state.contactList,toAdd]
            } else {
                updateContactList = [...state.contactList]
            }
            toUpdate = {
                ...state,
                contactList: updateContactList
            }
            break;
        }
            
        case 'updateContact': {
            const updateList = [...state.contactList];
            const toUpdateContact = update.data.toUpdate;
            const index = updateList.findIndex((item) => {
                return (item.recordID == toUpdateContact.recordID);
            });
            updateList[index] = toUpdateContact;
            toUpdate = {
                ...state,
                contactList: updateList
            }
            break;
        }
            
        case 'triggerSync': {
            AsyncStorage.setItem("@syncEnabled","true");
            toUpdate = {
                ...state,
                syncEnabled: true
            };
            break;
        }
        case 'syncResult': {
            const contactList = update.data.contactList;
            AsyncStorage.setItem("@contacts",JSON.stringify(contactList));
            toUpdate = {
                ...state,
                contactList: contactList
            } 
            break;
        }
    }
    return toUpdate;
}

export function UserController(props) {
    const [userState, updateUser] = useReducer(updater, initialState);
    const value = useMemo(() => [userState, updateUser], [userState]);
    return (<UserContext.Provider value={value}>
        {props.children}
    </UserContext.Provider>);
  }
  
  UserController.propTypes = {
      children: PropTypes.oneOfType([
          PropTypes.arrayOf(PropTypes.node),
          PropTypes.node
      ]).isRequired
  };
  