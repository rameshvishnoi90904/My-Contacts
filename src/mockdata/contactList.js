export const contactList = [
    {
      recordID: '1',
      givenName: 'Elon',
      familyName: 'Musk',
      emailAddresses: [{
        email: 'andre@gmail.com', 
        label: 'work'}],
      phoneNumbers: [{
        number: '+23493023902',
        label: 'personal'}],
      picture: '',
    },
    {
      recordID: '2',
      givenName: 'Jack',
      familyName: 'Ma',
      emailAddresses: [{
        email: 'andre@gmail.com', 
        label: 'work'}],
      phoneNumbers: [{
        number: '+23493023902',
        label: 'personal'}],
      picture: '',
    },
    {
      recordID: '3',
      givenName: 'Larry',
      familyName: 'Page',
      emailAddresses: [{
        email: 'andre@gmail.com', 
        label: 'work'}],
      phoneNumbers: [{
        number: '+23493023902',
        label: 'personal'}],
      picture: '',
    },
    {
      recordID: '4',
      givenName: 'Mark',
      familyName: 'Zuck',
      emailAddresses: [{
        email: 'andre@gmail.com', 
        label: 'work'}],
      phoneNumbers: [{
        number: '+23493023902',
        label: 'personal'}],
      picture: '',
    },
    {
      recordID: '5',
      givenName: 'Jeff',
      familyName: 'Bezos',
      emailAddresses: [{
        email: 'andre@gmail.com', 
        label: 'work'}],
      phoneNumbers: [{
        number: '+23493023902',
        label: 'personal'}],
      picture: '',
    },
];