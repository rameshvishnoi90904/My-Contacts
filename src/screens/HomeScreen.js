import React, {useEffect, useState, useContext} from 'react';
import {
  SafeAreaView,
  ScrollView,
  StatusBar,
  StyleSheet,
  Text,
  useColorScheme,
  View,
  Button,
  PermissionsAndroid,
  FlatList,
  Touchable,
  TouchableOpacity
} from 'react-native';

import Contacts from 'react-native-contacts';
import {checkMultiple, PERMISSIONS, requestMultiple, request} from 'react-native-permissions';
import AsyncStorage from '@react-native-async-storage/async-storage';
import Icon from 'react-native-vector-icons/MaterialCommunityIcons'
import { UserContext } from '../context/context';
import { getUserInitials } from '../helper';
const colorList = [
  '#4a5568',
  '#c53030',
  '#c05621',
  '#b7791f',
  '#2f855a',
  '#2c7a7b',
  '#2b6cb0',
  '#4c51bf',
  '#6b46c1',
  '#b83280',
]

const HomeScreen = ({navigation}) => {
    const [userState, updateState] = useContext(UserContext);
    const sync = () => {
        try {
          checkMultiple([PERMISSIONS.ANDROID.WRITE_CONTACTS, PERMISSIONS.ANDROID.READ_CONTACTS]).then((result) => {
            requestMultiple([PERMISSIONS.ANDROID.WRITE_CONTACTS, PERMISSIONS.ANDROID.READ_CONTACTS]).then((res) => {
              getContacts();
            });
          });
        } catch (error) {
            console.log("Error :: ",error)
        }
    }

    const getContacts = () => {
      Contacts.getAll().then(contacts => {
        const stringifiedValue = JSON.stringify(contacts);
        AsyncStorage.setItem("@contacts",stringifiedValue)
        updateState({type: "triggerSync"})
        updateState({type: "init", data: {contactList: contacts}})
      })
    }

    const navigateToContactDetail = (data, selectedColor) => {
      navigation.navigate("Contact Details",{
        contactInfo: data,
        selectedColor: selectedColor
      })
    }
    const navigateToEdit = (data) => {
      navigation.navigate("Edit Contact",{
        contactInfo: data,
      })
    }
    const renderContactItem = ({item, index}) => {
      let contactNumber;
      if (item.phoneNumbers && item.phoneNumbers[0]) {
        contactNumber = item.phoneNumbers[0].number;
      }
      let userInitials = getUserInitials(item);
      const selectedColor = colorList[index%10];
      return (
        <TouchableOpacity
        accessibilityRole="link"
        accessibilityHint={`${item.givenName} ${item.familyName} Contact Item`}
        style={styles.contactItem} 
        onPress={() => navigateToContactDetail(item, selectedColor)}>
          <View style={[styles.profilePicture, {backgroundColor: selectedColor}]}>
            <Text style={styles.userInitials}>{userInitials}</Text>
          </View>
          <View style={styles.contactInfo}>
            <View style={{flex: 1}}>
              <Text style={styles.contactName}>{`${item.givenName} ${item.familyName}`}</Text>
              <Text style={styles.contactNumber}>{contactNumber}</Text>
            </View>
              <TouchableOpacity onPress={() => navigateToEdit(item)}>
                <Icon name="account-edit" size={32}/>
              </TouchableOpacity>
          </View>
        </TouchableOpacity>
      )
    }

    let syncInfo;
    if (userState.contactList.length == 0) {
      syncInfo = (
        <View style={styles.syncContainer}>
          <Icon name="database-sync" size={52}/>
          <Text 
            accessibilityRole="header"
            accessibilityAutoFocus
            style={styles.syncText}>Sync now to import contact from your phone</Text>
        </View>
      )
    }

    let syncButton;
    if (!userState.syncEnabled || userState.contactList.length == 0) {
        syncButton = (
            <Button 
                testID="sync-btn"
                title="Sync" 
                accessible={true}
                accessibilityLabel="Sync contacts"
                accessibilityHint="Syncs contacts from device"
                onPress={sync}/>
        )
    }

    let contactList;
    if (userState.contactList.length > 0) {
        contactList = (
            <FlatList
                  testID="flatlist"
                  data={userState.contactList}
                  renderItem={renderContactItem}
                  keyExtractor={(item) => item.recordID}
                ></FlatList>
        )
    }

    return (
        <View style={styles.container} testID="root">
            {contactList}
            {syncInfo}
            {syncButton}
        </View>
    )
}
const styles = StyleSheet.create({
    container: {
      flex: 1,
      paddingHorizontal: 16,
      paddingVertical: 12,
      backgroundColor: 'white'
    },
    contactItem: {
      marginBottom: 12,
      flexDirection: 'row',
    },
    profilePicture: {
      width: 60,
      height: 60,
      borderRadius: 30,
      marginRight: 16,
      backgroundColor: '#8D8D8D',
      alignItems: 'center',
      justifyContent: 'center'
    },
    contactInfo: {
      borderBottomWidth: 1,
      flex: 1,
      borderBottomColor: '#C9C9C9',
      flexDirection: 'row',
      alignItems: 'center'
    },
    contactName: {
      fontSize: 16,
      color: '#333333'
    },
    contactNumber: {
      color: '#8D8D8D'
    },
    userInitials: {
      color: 'white',
      fontSize: 22,
      fontWeight: 'bold'
    },
    syncText: {
      fontSize: 18
    },
    syncContainer: {
      flex: 1, 
      alignItems: 'center', 
      justifyContent: 'center'
    }
  });

export default HomeScreen;