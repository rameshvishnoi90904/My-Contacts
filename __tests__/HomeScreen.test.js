import { expect } from "@jest/globals";
import React from "react";
import renderer from "react-test-renderer";
import HomeScreen from '../src/screens/HomeScreen';
import {contactList} from '../src/mockData/contactList';
import {UserController, UserContext} from '../src/context/context';
import { render } from "@testing-library/react-native";

describe("<HomeScreen />", () => {
    it('renders correctly across screens', () => {
        const container = renderer.create(
        <UserController value={{contactList: contactList}}>
          <HomeScreen />
        </UserController>
        )
        const tree = container.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('contact list match',() => {
      const component = render(
        <UserContext.Provider value={[{contactList: contactList},() => {}]}>
          <HomeScreen />
        </UserContext.Provider>
      )
      const flatList = component.getByTestId("flatlist");
      expect(flatList.props.data.length).toBe(contactList.length)
    })
});