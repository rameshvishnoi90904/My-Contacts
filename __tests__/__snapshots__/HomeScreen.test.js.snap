// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomeScreen /> renders correctly across screens 1`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "flex": 1,
      "paddingHorizontal": 16,
      "paddingVertical": 12,
    }
  }
  testID="root"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <Text
      allowFontScaling={false}
      style={
        Array [
          Object {
            "color": undefined,
            "fontSize": 52,
          },
          undefined,
          Object {
            "fontFamily": "Material Design Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      ó°³¿
    </Text>
    <Text
      accessibilityAutoFocus={true}
      accessibilityRole="header"
      style={
        Object {
          "fontSize": 18,
        }
      }
    >
      Sync now to import contact from your phone
    </Text>
  </View>
  <View
    accessibilityLabel="Sync contacts"
    accessibilityRole="button"
    accessible={true}
    collapsable={false}
    focusable={true}
    nativeID="animatedComponent"
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
    testID="sync-btn"
  >
    <View
      style={
        Array [
          Object {},
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Sync
      </Text>
    </View>
  </View>
</View>
`;
